patch/kernel/rk3399-legacy/pci-host-rockchip-0000-scan-bus-fix-from-rockchip64.patch

```diff
@@ -0,0 +1,289 @@
++}
++
+ static int rockchip_pcie_rd_other_conf(struct rockchip_pcie *rockchip,
+ 				       struct pci_bus *bus, u32 devfn,
+ 				       int where, int size, u32 *val)
+@@ -385,11 +418,23 @@ static int rockchip_pcie_rd_other_conf(struct rockchip_pcie *rockchip,
+ 	busdev = PCIE_ECAM_ADDR(bus->number, PCI_SLOT(devfn),
+ 				PCI_FUNC(devfn), where);
+ 
++	if (bus->number > 0x1f) {
++		*val = 0;
++		return PCIBIOS_DEVICE_NOT_FOUND;
++	}
++
+ 	if (!IS_ALIGNED(busdev, size)) {
+ 		*val = 0;
+ 		return PCIBIOS_BAD_REGISTER_NUMBER;
+ 	}
+ 
++	if (bus->parent->number == rockchip->root_bus_nr)
++		rockchip_pcie_cfg_configuration_accesses(rockchip,
++						AXI_WRAPPER_TYPE0_CFG);
++	else
++		rockchip_pcie_cfg_configuration_accesses(rockchip,
++						AXI_WRAPPER_TYPE1_CFG);
++
+ 	if (size == 4) {
+ 		*val = readl(rockchip->reg_base + busdev);
+ 	} else if (size == 2) {
+@@ -411,9 +456,19 @@ static int rockchip_pcie_wr_other_conf(struct rockchip_pcie *rockchip,
+ 
+ 	busdev = PCIE_ECAM_ADDR(bus->number, PCI_SLOT(devfn),
+ 				PCI_FUNC(devfn), where);
++
++	if (bus->number > 0x1f)
++		return PCIBIOS_DEVICE_NOT_FOUND;
+ 	if (!IS_ALIGNED(busdev, size))
+ 		return PCIBIOS_BAD_REGISTER_NUMBER;
+ 
++	if (bus->parent->number == rockchip->root_bus_nr)
++		rockchip_pcie_cfg_configuration_accesses(rockchip,
++						AXI_WRAPPER_TYPE0_CFG);
++	else
++		rockchip_pcie_cfg_configuration_accesses(rockchip,
++						AXI_WRAPPER_TYPE1_CFG);
++
+ 	if (size == 4)
+ 		writel(val, rockchip->reg_base + busdev);
+ 	else if (size == 2)
+@@ -638,6 +693,11 @@ static int rockchip_pcie_init_port(struct rockchip_pcie *rockchip)
+ 	status |= PCI_EXP_LNKCTL_CCC;
+ 	rockchip_pcie_write(rockchip, status, PCIE_RC_CONFIG_LCS);
+ 
++	/* Set RC's RCB to 128 */
++	status = rockchip_pcie_read(rockchip, PCIE_RC_CONFIG_LCS);
++	status |= PCI_EXP_LNKCTL_RCB;
++	rockchip_pcie_write(rockchip, status, PCIE_RC_CONFIG_LCS);
++
+ 	/* Enable Gen1 training */
+ 	rockchip_pcie_write(rockchip, PCIE_CLIENT_LINK_TRAIN_ENABLE,
+ 			    PCIE_CLIENT_CONFIG);
+@@ -696,15 +756,8 @@ static int rockchip_pcie_init_port(struct rockchip_pcie *rockchip)
+ 		rockchip_pcie_write(rockchip, status, PCIE_RC_CONFIG_LINK_CAP);
+ 	}
+ 
+-	rockchip_pcie_write(rockchip, 0x0, PCIE_RC_BAR_CONF);
+-
+-	rockchip_pcie_write(rockchip,
+-			    (RC_REGION_0_ADDR_TRANS_L + RC_REGION_0_PASS_BITS),
+-			    PCIE_CORE_OB_REGION_ADDR0);
+-	rockchip_pcie_write(rockchip, RC_REGION_0_ADDR_TRANS_H,
+-			    PCIE_CORE_OB_REGION_ADDR1);
+-	rockchip_pcie_write(rockchip, 0x0080000a, PCIE_CORE_OB_REGION_DESC0);
+-	rockchip_pcie_write(rockchip, 0x0, PCIE_CORE_OB_REGION_DESC1);
++	rockchip_pcie_cfg_configuration_accesses(rockchip,
++					AXI_WRAPPER_TYPE0_CFG);
+ 
+ 	return 0;
+ }
+@@ -980,9 +1033,9 @@ static int rockchip_pcie_parse_dt(struct rockchip_pcie *rockchip)
+ 		return PTR_ERR(rockchip->aclk_rst);
+ 	}
+ 
+-	rockchip->ep_gpio = devm_gpiod_get_optional(dev, "ep", GPIOD_OUT_HIGH);
++	rockchip->ep_gpio = devm_gpiod_get(dev, "ep", GPIOD_OUT_HIGH);
```

Comment:
This change broke `legacy` for devices that do not have `ep-gpios` node defined in device tree - including NanoPiM4/M4V2 ;-).

I pushed some changes to the whole patch as it also removed some unrelated fixes that were commited in FriendlyARM's fork.